homeassistant/components/smarttub/controller.py

```diff
@@ -0,0 +1,115 @@
+"""Interface to the SmartTub API."""
+
+from datetime import timedelta
+import logging
+
+import async_timeout
+from smarttub import APIError, LoginFailed, SmartTub
+from smarttub.api import Account
+
+from homeassistant.config_entries import SOURCE_REAUTH
+from homeassistant.const import CONF_EMAIL, CONF_PASSWORD
+from homeassistant.helpers import device_registry as dr
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN, POLLING_TIMEOUT, SCAN_INTERVAL
+from .helpers import get_spa_name
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class SmartTubController:
+    """Interface between Home Assistant and the SmartTub API."""
+
+    def __init__(self, hass):
+        """Initialize an interface to SmartTub."""
+        self._hass = hass
+        self._api = SmartTub(async_get_clientsession(hass))
+        self._account = None
+        self.spas = set()
+        self._spa_devices = {}
+
+        self.coordinator = None
+
+    async def async_setup_entry(self, entry):
+        """Perform initial setup.
+
+        Authenticate, query static state, set up polling, and otherwise make
+        ready for normal operations .
+        """
+
+        self._account = await self.login(
+            entry.data[CONF_EMAIL], entry.data[CONF_PASSWORD]
+        )
+        if self._account is None:
+            # credentials were changed or invalidated, we need new ones
+            self._hass.async_create_task(
+                self._hass.config_entries.flow.async_init(
+                    DOMAIN,
+                    context={"source": SOURCE_REAUTH},
+                    data=None,
+                )
+            )
+
+            return False
```

Comment:
When this returns `False` it will raise `ConfigEntryNotReady` which means its going to try to login again with credentials we know are invalid which may get the user locked out.


The `async_setup_entry` in `__init__` should return False when we know authentication creds are bad and then start the reauth flow.
We only want to raise `ConfigEntryNotReady` when the service is offline is unreachable.

Reply:
OK, we now return `False` if the credentials are bad, and raise `ConfigEntryNotReady` on failure to connect etc.